File: src/main/java/com/teamwizardry/librarianlib/math/shapes/Arc2D.java
package com.teamwizardry.librarianlib.math.shapes;

import com.teamwizardry.librarianlib.math.Vec2d;

import java.util.ArrayList;

/**
 * Created by Saad on 12/7/2016.
 */
public class Arc2D implements IShape2D {

    /**
     * Height of the arc in blocks
     */
    private float height = 2;

    /**
     * Particles per arc
     */
    private int particles = 100;

    /**
     * The two points the arc will connect from and to
     */
    private Vec2d origin, target;

    public Arc2D(Vec2d origin, Vec2d target, float height, int particleCount) {
        this.height = height;
        this.particles = particleCount;
        this.origin = origin;
        this.target = target;
    }

    /**
     * Will return a list of points in order that define every point of the arc
     *
     * @return Will return the list of points required
     */
    @Override
    public ArrayList<Vec2d> getPoints() {
        ArrayList<Vec2d> locs = new ArrayList<>();
        Vec2d link = target.sub(origin);
        float length = (float) link.length();
        float pitch = (float) (4 * height / Math.pow(length, 2));
        for (int i = 0; i < particles; i++) {
            Vec2d tmp = new Vec2d(link.x, link.y).normalize();
            Vec2d v = new Vec2d(tmp.x * length * i / particles, tmp.y * length * i / particles);
            float x = ((float) i / particles) * length - length / 2;
            float y = (float) (-pitch * Math.pow(x, 2) + height);
            locs.add(origin.add(v).add(new Vec2d(0, y)));
        }
        return locs;
    }
}

File: src/main/java/com/teamwizardry/librarianlib/math/shapes/Arc2D.kt
package com.teamwizardry.librarianlib.math.shapes

import com.teamwizardry.librarianlib.math.Vec2d

import java.util.ArrayList

/**
 * Created by Saad on 12/7/2016.
 */
class Arc2D @JvmOverloads constructor(val origin: Vec2d, val target: Vec2d, val height: Float = 2f, val particles: Int = 100) : IShape2D {

    /**
     * Will return a list of points in order that define every point of the arc

     * @return Will return the list of points required
     */
    override val points: ArrayList<Vec2d>
        get() {
            val locs = ArrayList<Vec2d>()
            val link = target.sub(origin)
            val length = link.length().toFloat()
            val pitch = (4 * height / Math.pow(length.toDouble(), 2.0)).toFloat()
            for (i in 0..particles - 1) {
                val tmp = Vec2d(link.x, link.y).normalize()
                val v = Vec2d(tmp.x * length.toDouble() * i.toDouble() / particles, tmp.y * length.toDouble() * i.toDouble() / particles)
                val x = i.toFloat() / particles * length - length / 2
                val y = (-pitch * Math.pow(x.toDouble(), 2.0) + height).toFloat()
                locs.add(origin.add(v).add(Vec2d(0.0, y.toDouble())))
            }
            return locs
        }
}